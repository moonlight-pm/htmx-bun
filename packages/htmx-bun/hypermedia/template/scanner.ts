import { voids } from "./ast";

export enum TokenType {
    Whitespace = "Whitespace",
    TagName = "TagName",
    AttributeName = "AttributeName",
    OpenAngleBracket = "OpenAngleBracket",
    CloseAngleBracket = "CloseAngleBracket",
    Slash = "Slash",
    Equal = "Equal",
    Expression = "Expression",
    Text = "Text",
}

export interface Token {
    type: TokenType;
    position: number;
    value: string;
}

export function scanPartial(source: string): Token[] {
    const scanner = new Scanner();
    return scanner.scan(source);
}

export function htmlStartIndex(source: string) {
    return source.search(/^<\w+/m);
}

class Scanner {
    source = "";
    position = 0;
    tokens: Token[] = [];

    scan(source: string): Token[] {
        this.source = source;
        this.position = htmlStartIndex(source);
        if (this.position === -1) {
            return [];
        }
        while (this.position < this.source.length) {
            const start = this.position;
            this.matchNext();
            if (start === this.position) {
                this.error("Unexpected character");
            }
        }
        return this.tokens;
    }

    token(type: TokenType, position: number) {
        const token = {
            type,
            position,
            value: this.source.slice(position, this.position),
        };
        this.tokens.push(token);
    }

    get ultimate(): Token | undefined {
        return this.tokens[this.tokens.length - 1];
    }

    private matchNext() {
        this.matchElement();
        this.matchText();
    }

    private matchElement() {
        if (
            this.peekStringMatch(2, "</") ||
            !this.matchCharacterType(TokenType.OpenAngleBracket, "<")
        ) {
            return;
        }
        if (!this.matchTagName()) {
            this.error("Expected tag name");
        }
        const tag = this.ultimate!.value;
        this.matchWhitespace();
        this.matchAttributes();
        let selfClosing = false;
        if (this.matchCharacterType(TokenType.Slash, "/")) {
            selfClosing = true;
        }
        if (!this.matchCharacterType(TokenType.CloseAngleBracket, ">")) {
            this.error("Expected closing brace");
        }
        if (voids.includes(tag) || selfClosing) {
            return;
        }
        this.matchContent();
        if (!this.matchCloseTag(tag)) {
            this.error(`Expected closing tag for ${tag}`);
        }
    }

    private matchCloseTag(tag: string) {
        return (
            this.matchCharacterType(TokenType.OpenAngleBracket, "<") &&
            this.matchCharacterType(TokenType.Slash, "/") &&
            this.matchTagName() &&
            this.ultimate!.value === tag &&
            this.matchCharacterType(TokenType.CloseAngleBracket, ">")
        );
    }

    private matchTagName() {
        const position = this.position;
        const match = this.source
            .slice(this.position)
            .match(/^[a-z][a-z0-9]*(-[a-z]+)*/);
        if (match) {
            this.position += match[0].length;
            this.token(TokenType.TagName, position);
            return true;
        }
    }

    private matchAttributes() {
        while (!this.peekCharacter("/>")) {
            this.matchAttribute();
            this.matchWhitespace();
        }
    }

    private matchAttribute() {
        if (!this.matchAttributeName()) {
            this.error("Expected attribute name");
        }
        if (!this.matchCharacterType(TokenType.Equal, "=")) {
            return;
        }
        if (!this.matchAttributeValue()) {
            this.error("Expected attribute value");
        }
    }

    private matchAttributeName() {
        const position = this.position;
        const match = this.source
            .slice(this.position)
            .match(/^[a-z]+((:|::|-)[a-z]+)*/);
        if (match) {
            this.position += match[0].length;
            this.token(TokenType.AttributeName, position);
            return true;
        }
    }

    private matchAttributeValue() {
        if (this.matchExpression()) {
            return true;
        }
        if (!this.peekCharacter('"')) {
            return;
        }
        this.position++;
        const position = this.position;
        this.matchAttributeText();
        if (!this.peekCharacter('"')) {
            this.error("Expected closing quote");
            return;
        }
        this.position++;
        return true;
    }

    private matchAttributeText(): true | undefined {
        const position = this.position;
        while (
            !this.peekCharacter('{"') &&
            this.position < this.source.length
        ) {
            this.position++;
        }
        if (position !== this.position) {
            this.token(TokenType.Text, position);
        }
        if (this.peekCharacter("{")) {
            this.matchExpression();
            return this.matchAttributeText();
        }
        if (position === this.position) {
            return;
        }
        return true;
    }

    private matchContent() {
        while (
            this.position < this.source.length &&
            !this.peekStringMatch(2, "</")
        ) {
            this.matchText();
            this.matchElement();
        }
    }

    private matchText(): true | undefined {
        const position = this.position;
        while (
            !this.peekCharacter("{<") &&
            this.position < this.source.length
        ) {
            this.position++;
        }
        if (position !== this.position) {
            this.token(TokenType.Text, position);
        }
        if (this.peekCharacter("{")) {
            this.matchExpression();
            return this.matchText();
        }
        if (this.peekCharacter("<") && !this.peekStringMatch(2, "</")) {
            this.matchElement();
            return this.matchText();
        }
        if (position === this.position) {
            return;
        }
        return true;
    }

    private matchExpression() {
        const position = this.position;
        if (this.source[this.position] !== "{") {
            return;
        }
        this.position++;
        while (this.source[this.position] !== "}") {
            this.position++;
        }
        this.position++;
        this.token(TokenType.Expression, position);
        return true;
    }

    private matchWhitespace() {
        const position = this.position;
        const match = this.source.slice(this.position).match(/^\s+/);
        if (!match) {
            return;
        }
        this.position += match[0].length;
        this.token(TokenType.Whitespace, position);
        return true;
    }

    private matchCharacterType(type: TokenType, chars: string) {
        const position = this.position;
        if (!chars.split("").includes(this.source[this.position])) {
            return;
        }
        this.position++;
        this.token(type, position);
        return true;
    }

    private peekCharacter(chars: string) {
        if (chars.split("").includes(this.source[this.position])) {
            return this.source[this.position];
        }
    }

    private peekStringMatch(length: number, str: string) {
        return this.source.slice(this.position, this.position + length) === str;
    }

    private error(message: string) {
        const line = this.source.slice(0, this.position).split("\n").length;
        const column =
            this.position - this.source.lastIndexOf("\n", this.position);
        throw new Error(
            `Syntax Error: ${message} at line ${line}, column ${column}: ${
                this.source[this.position]
            }\n`,
        );
    }
}
