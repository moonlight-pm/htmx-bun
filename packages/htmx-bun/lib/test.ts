import { afterAll } from "bun:test";
import { mkdtempSync, rmdirSync } from "node:fs";
import { tmpdir } from "node:os";

export function makeTemporaryDirectory() {
    const dir = mkdtempSync(`${tmpdir()}/htmx-bun-`);
    afterAll(() => {
        rmdirSync(dir, { recursive: true });
    });
    return dir;
}
