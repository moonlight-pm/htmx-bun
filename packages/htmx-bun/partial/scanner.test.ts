import { expect, test } from "bun:test";
import { scanPartial } from "./scanner";

const source1 = `
<item id="1" class={item.class}>{item.name}</item>
`;
test("scanner", () => {
    const tokens = scanPartial(source1);
    expect(tokens).toMatchSnapshot();
});
