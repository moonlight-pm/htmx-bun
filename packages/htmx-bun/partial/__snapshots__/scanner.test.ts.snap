// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`scanner 1`] = `
[
  {
    "position": 1,
    "type": "OpenAngleBracket",
    "value": "<",
  },
  {
    "position": 2,
    "type": "Identifier",
    "value": "item",
  },
  {
    "position": 6,
    "type": "Whitespace",
    "value": " ",
  },
  {
    "position": 7,
    "type": "Identifier",
    "value": "id",
  },
  {
    "position": 9,
    "type": "Equal",
    "value": "=",
  },
  {
    "position": 10,
    "type": "String",
    "value": ""1"",
  },
  {
    "position": 13,
    "type": "Whitespace",
    "value": " ",
  },
  {
    "position": 14,
    "type": "Identifier",
    "value": "class",
  },
  {
    "position": 19,
    "type": "Equal",
    "value": "=",
  },
  {
    "position": 20,
    "type": "Expression",
    "value": "{item.class}",
  },
  {
    "position": 32,
    "type": "CloseAngleBracket",
    "value": ">",
  },
  {
    "position": 33,
    "type": "Expression",
    "value": "{item.name}",
  },
  {
    "position": 44,
    "type": "OpenAngleBracket",
    "value": "<",
  },
  {
    "position": 45,
    "type": "Slash",
    "value": "/",
  },
  {
    "position": 46,
    "type": "Identifier",
    "value": "item",
  },
  {
    "position": 50,
    "type": "CloseAngleBracket",
    "value": ">",
  },
  {
    "position": 51,
    "type": "Text",
    "value": 
"
"
,
  },
]
`;
